x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/GetApiKeyRequest.java

```diff
@@ -0,0 +1,146 @@
+            @Nullable String apiKeyName) {
+        this.realmName = realmName;
+        this.userName = userName;
+        this.apiKeyId = apiKeyId;
+        this.apiKeyName = apiKeyName;
+    }
+
+    public String getRealmName() {
+        return realmName;
+    }
+
+    public String getUserName() {
+        return userName;
+    }
+
+    public String getApiKeyId() {
+        return apiKeyId;
+    }
+
+    public String getApiKeyName() {
+        return apiKeyName;
+    }
+
+    /**
+     * Creates get api key request for given realm name
+     * @param realmName realm name
+     * @return {@link GetApiKeyRequest}
+     */
+    public static GetApiKeyRequest usingRealmName(String realmName) {
+        return new GetApiKeyRequest(realmName, null, null, null);
+    }
+
+    /**
+     * Creates get API key request for given user name
+     * @param userName user name
+     * @return {@link GetApiKeyRequest}
+     */
+    public static GetApiKeyRequest usingUserName(String userName) {
+        return new GetApiKeyRequest(null, userName, null, null);
+    }
+
+    /**
+     * Creates get API key request for given realm and user name
+     * @param realmName realm name
+     * @param userName user name
+     * @return {@link GetApiKeyRequest}
+     */
+    public static GetApiKeyRequest usingRealmAndUserName(String realmName, String userName) {
+        return new GetApiKeyRequest(realmName, userName, null, null);
+    }
+
+    /**
+     * Creates get API key request for given api key id
+     * @param apiKeyId api key id
+     * @return {@link GetApiKeyRequest}
+     */
+    public static GetApiKeyRequest usingApiKeyId(String apiKeyId) {
+        return new GetApiKeyRequest(null, null, apiKeyId, null);
+    }
+
+    /**
+     * Creates get api key request for given api key name
+     * @param apiKeyName api key name
+     * @return {@link GetApiKeyRequest}
+     */
+    public static GetApiKeyRequest usingApiKeyName(String apiKeyName) {
+        return new GetApiKeyRequest(null, null, null, apiKeyName);
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException validationException = null;
+        if (Strings.hasText(realmName) == false && Strings.hasText(userName) == false && Strings.hasText(apiKeyId) == false
+                && Strings.hasText(apiKeyName) == false) {
+            validationException = addValidationError("One of [api key id, api key name, username, realm name] must be specified", null);
+        }
+        if (Strings.hasText(realmName) || Strings.hasText(userName)) {
+            if (Strings.hasText(apiKeyId)) {
+                validationException = addValidationError("api key id must not be specified when username or realm name is specified", null);
+            }
+            if (Strings.hasText(apiKeyName)) {
+                validationException = addValidationError("api key name must not be specified when username or realm name is specified",
```

Comment:
Hmm we may want to revisit this. Api key names should be unique per user So it makes since to send a user, realm, and key name. I guess in that case we could always send an id. Just something to think about

Reply:
I think we both suggest to have the key name by unique in the entire index. What Jay is suggesting here is that while it doesn't make much sense to add the realm and username in a request where the name or id is specified, we should think about allowing it. This would meant that they need to get the user and realm correct in addition to the api key name / api key id in the request. 

I'm fine either way I think but I would prefer that if we keep the current behavior we flip the order of validation and add a validation error to the effect of 

> "user name and/or realm name should not be specified when the api key name is specified" 

and 

> "user name and/or realm name should not be specified when the api key id is specified" 

and do the same for #37628